spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    this.startSnapshotId = options.get("starting-snapshot-id").map(Long::parseLong).orElse(null);
+    if (startSnapshotId != null) {
+      if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), startSnapshotId)) {
+        throw new IllegalStateException("The option starting-snapshot-id " + startSnapshotId +
+          "is not a valid snapshot id");
+      }
+    }
+
+    this.splitSize = options.get("split-size").map(Long::parseLong).orElse(
+        Optional.ofNullable(table.properties().get(TableProperties.SPLIT_SIZE))
+            .map(Long::parseLong)
+            .orElse(TableProperties.SPLIT_SIZE_DEFAULT));
+    this.lookback = options.get("lookback").map(Integer::parseInt).orElse(
+        Optional.ofNullable(table.properties().get(TableProperties.SPLIT_LOOKBACK))
+            .map(Integer::parseInt)
+            .orElse(TableProperties.SPLIT_LOOKBACK_DEFAULT));
+    this.openFileCost = options.get("file-open-cost").map(Long::parseLong).orElse(
+        Optional.ofNullable(table.properties().get(TableProperties.SPLIT_OPEN_FILE_COST))
+            .map(Long::parseLong)
+            .orElse(TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT));
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    table.refresh();
+
+    if (start.isPresent() && !StreamingOffset.START_OFFSET.equals(start.get())) {
+      this.startOffset = (StreamingOffset) start.get();
+      this.endOffset = (StreamingOffset) end.orElse(calculateEndOffset(startOffset));
+    } else {
+      // If starting offset is "START_OFFSET" (there's no snapshot in the last batch), or starting
+      // offset is not set, then we need to calculate the starting offset again.
+      this.startOffset = calculateStartingOffset();
+      this.endOffset = calculateEndOffset(startOffset);
+    }
+  }
+
+  @Override
+  public Offset getStartOffset() {
+    if (startOffset == null) {
+      throw new IllegalStateException("Start offset is not set");
+    }
+
+    return startOffset;
+  }
+
+  @Override
+  public Offset getEndOffset() {
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
+    }
+
+    return endOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+    // Since all the data and metadata of Iceberg is as it is, nothing needs to commit when
+    // offset is processed, so no need to implement this method.
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected List<CombinedScanTask> tasks() {
+    LOG.info("Processing data from {} exclusive to {} inclusive", startOffset, endOffset);
```

Comment:
Minor: Can you move this to after the check for whether there is work to do?

It would also be good to add an info message if there are no tasks that says there are no tasks.